static/app/components/codeSnippet.tsx

```diff
@@ -1,13 +1,45 @@
-import {useEffect, useRef, useState} from 'react';
+import {ComponentType, useEffect, useRef, useState} from 'react';
+import {createPortal} from 'react-dom';
 import styled from '@emotion/styled';
 import Prism from 'prismjs';
 
 import {Button} from 'sentry/components/button';
+import {NODE_ENV} from 'sentry/constants';
 import {IconCopy} from 'sentry/icons';
 import {t} from 'sentry/locale';
 import {space} from 'sentry/styles/space';
 import {loadPrismLanguage} from 'sentry/utils/loadPrismLanguage';
 
+/**
+ * Replaces tokens in a DOM element with a span element.
+ * @param element DOM element in which the tokens will be replaced
+ * @param tokens array of tokens to be replaced
+ * @returns object with keys as tokens and values as array of HTMLSpanElement
+ */
+export function replaceTokensWithSpan(element: HTMLElement, tokens: string[]) {
+  const replaceRegex = new RegExp(`(${tokens.join('|')})`, 'g');
+  element.innerHTML = element.innerHTML.replace(
+    replaceRegex,
+    '<span data-token="$1"></span>'
+  );
+
+  const nodes = tokens.reduce(
+    (acc, token) => {
+      const tokenNodes = Array.from<HTMLSpanElement>(
+        element.querySelectorAll(`[data-token="${token}"]`)
+      );
+      return tokenNodes.length > 0
+        ? {
+            ...acc,
+            [token]: tokenNodes,
+          }
+        : acc;
+    },
+    {} as Record<string, HTMLSpanElement[]>
+  );
+  return nodes;
+}
+
```

Comment:
I'm a bit concerned about iterating over and replacing component children like this — it's error-prone and can cause flashes/unexpected behaviors. I'm not sure how we can better implement this given the requirements. @evanpurkhiser Any ideas?

Reply:
Maybe we can decompose `<CodeSnippet />` into sub-snippets, like:
```jsx
<CodeSnippet>
  <CodeSnippet.SubSnippet>
      …
  </CodeSnippet.SubSnippet>
  <Token>___ORG_AUTH_TOKEN___</Token>
   <CodeSnippet.SubSnippet>
      …
  </CodeSnippet.SubSnippet>
</CodeSnippet>
```

And then we can only highlight the code inside `CodeSnippet.SubSnippet`. This way, we don't need the async `applyTokenReplacers()` callback, and can potentially solve the testing issues you mentioned above.

Reply:
Not sure if we can guarantee correct syntax highlighting for every language with that approach as the single sub-snippets are missing context 🤔

I could still try prism's low level API or react-prism-renderer. That would probably be the ideal approach but also a bigger change. 
Unless our homemade highlighter is ready to support all the languages, that we have in the onboarding.

Reply:
@vuluongj20 and me agreed that we will move this functionality into a wrapper component to not pollute the base component with very requirement specific code.